From 47eedc730bd8ea2c0e5f2cdc3cddda6f2938d1b9 Mon Sep 17 00:00:00 2001
From: Jevin Tang <a21266@motorola.com>
Date: Fri, 15 Jan 2016 13:02:45 -0600
Subject: [PATCH 621/959] IKSWM-12849 diag: provide peripherals query

User space can query the supported peripherals status

Change-Id: Iddc6410e887d6f9363b50bf8bf3060d47fcc7677
Signed-off-by: Jevin Tang <a21266@motorola.com>
Reviewed-on: http://gerrit.mot.com/813731
SLTApproved: Slta Waiver <sltawvr@motorola.com>
SME-Granted: SME Approvals Granted
Tested-by: Jira Key <jirakey@motorola.com>
Reviewed-by: Christopher Fries <cfries@motorola.com>
Submit-Approved: Jira Key <jirakey@motorola.com>
---
 drivers/char/diag/diagchar_core.c | 16 ++++++++++++++++
 drivers/char/diag/diagfwd_cntl.c  |  8 ++++++++
 include/linux/diagchar.h          |  1 +
 3 files changed, 25 insertions(+)

diff --git a/drivers/char/diag/diagchar_core.c b/drivers/char/diag/diagchar_core.c
index ffe84308de9..15b550e6fbd 100644
--- a/drivers/char/diag/diagchar_core.c
+++ b/drivers/char/diag/diagchar_core.c
@@ -161,6 +161,7 @@ uint16_t diag_debug_mask;
 void *diag_ipc_log;
 #endif
 
+extern uint16_t md_support;
 /*
  * Returns the next delayed rsp id. If wrapping is enabled,
  * wraps the delayed rsp id to DIAGPKT_MAX_DELAYED_RSP.
@@ -1546,6 +1547,15 @@ static int diag_ioctl_hdlc_toggle(unsigned long ioarg)
 	return 0;
 }
 
+static int diag_ioctl_md_support_list(unsigned long ioarg)
+{
+	if (copy_to_user((void __user *)ioarg, &md_support,
+			sizeof(md_support)))
+		return -EFAULT;
+	else
+		return 0;
+}
+
 static int diag_ioctl_register_callback(unsigned long ioarg)
 {
 	struct diag_callback_reg_t reg;
@@ -1778,6 +1788,9 @@ long diagchar_compat_ioctl(struct file *filp,
 	case DIAG_IOCTL_HDLC_TOGGLE:
 		result = diag_ioctl_hdlc_toggle(ioarg);
 		break;
+	case DIAG_IOCTL_MD_SUPPORT_LIST:
+		result = diag_ioctl_md_support_list(ioarg);
+		break;
 	}
 	return result;
 }
@@ -1896,6 +1909,9 @@ long diagchar_ioctl(struct file *filp,
 	case DIAG_IOCTL_HDLC_TOGGLE:
 		result = diag_ioctl_hdlc_toggle(ioarg);
 		break;
+	case DIAG_IOCTL_MD_SUPPORT_LIST:
+		result = diag_ioctl_md_support_list(ioarg);
+		break;
 	}
 	return result;
 }
diff --git a/drivers/char/diag/diagfwd_cntl.c b/drivers/char/diag/diagfwd_cntl.c
index efd74d9132d..8be4ff99205 100644
--- a/drivers/char/diag/diagfwd_cntl.c
+++ b/drivers/char/diag/diagfwd_cntl.c
@@ -25,6 +25,8 @@
 
 #define FEATURE_SUPPORTED(x)	((feature_mask << (i * 8)) & (1 << x))
 
+uint16_t md_support;
+
 /* tracks which peripheral is undergoing SSR */
 static uint16_t reg_dirty;
 static void diag_notify_md_client(uint8_t peripheral, int data);
@@ -109,6 +111,12 @@ void diag_notify_md_client(uint8_t peripheral, int data)
 	int stat = 0;
 	struct siginfo info;
 
+	if (data == DIAG_STATUS_OPEN) {
+		md_support |= PERIPHERAL_MASK(peripheral);
+	} else if (data == DIAG_STATUS_CLOSED) {
+		md_support ^= PERIPHERAL_MASK(peripheral);
+	}
+
 	if (driver->logging_mode != MEMORY_DEVICE_MODE)
 		return;
 
diff --git a/include/linux/diagchar.h b/include/linux/diagchar.h
index 4581f5046ff..ea7923894f7 100644
--- a/include/linux/diagchar.h
+++ b/include/linux/diagchar.h
@@ -65,6 +65,7 @@
 #define DIAG_IOCTL_PERIPHERAL_BUF_DRAIN		36
 #define DIAG_IOCTL_REGISTER_CALLBACK	37
 #define DIAG_IOCTL_HDLC_TOGGLE	38
+#define DIAG_IOCTL_MD_SUPPORT_LIST 39
 
 /* PC Tools IDs */
 #define APQ8060_TOOLS_ID	4062
-- 
2.11.0

